File: 백준(python)/baekjoon_17952.py
# https://www.acmicpc.net/problem/17952

import sys
from collections import deque

N = int(sys.stdin.readline().strip())
study = deque([])
res = 0

for _ in range(N):
    inList = list(map(int, sys.stdin.readline().strip().split()))
    
    # deque가 비어있는 경우
    if not study:
        if inList[0] == 1:
            if inList[2] - 1 == 0:
                res += inList[1]
            else:
                study.append((inList[1], inList[2]-1))
        # 과제목록이 비어있고 새로운 과제가 없을 경우 할 일이 없음을 의미
        else:
            pass
    
    # deque가 비어있지 않은 경우
    else:
        if inList[0] == 1:
            if inList[2] - 1 == 0:
                res += inList[1]
            else:
                study.append((inList[1], inList[2]-1))
            
        else:
            # 새로운 과제가 없을 경우 최근에 하던 과제를 진행
            nowStudy = study.pop()
            if nowStudy[1] - 1 ==0:
                res += nowStudy[0]
            else:
                study.append((nowStudy[0], nowStudy[1]-1))

print(res)

File: 백준/baekjoon_17952.java
package baekjoon;

// https://www.acmicpc.net/problem/17952

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class baekjoon_17952 {
	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		
		// 과제 클래스로 stack 생성
		Stack<Task> taskStack = new Stack<>();
		int res = 0;
		
		for(int i = 0; i < N; i++) {
			String[] inTask = br.readLine().split(" ");
			
			if(inTask[0].equals("1")) {
				if(Integer.parseInt(inTask[2]) - 1 == 0) {
					res += Integer.parseInt(inTask[1]);
				}
				else {
					taskStack.add(new Task(Integer.parseInt(inTask[1]), Integer.parseInt(inTask[2])-1));
				}
			}
			else {
				if(taskStack.empty()) {
					continue;
				}
				else {
					Task nowTask = taskStack.pop();
					if(nowTask.time - 1 == 0) {
						res += nowTask.score;
					}
					else {
						taskStack.add(new Task(nowTask.score, nowTask.time - 1));
					}
				}			
			}
		}
		
		System.out.println(res);
	}
}

// 과제 클래스 생성
class Task{
	int score;
	int time;
	
	Task(int score, int time){
		this.score = score;
		this.time = time;
	}
}
